src/main/java/de/adorsys/keycloak/config/repository/ClientRepository.java

```diff
@@ -56,11 +56,29 @@ public ClientRepository(RealmRepository realmRepository) {
     }
 
     public Optional<ClientRepresentation> searchByClientId(String realmName, String clientId) {
-        return Optional.ofNullable(clientId)
-                .filter(id -> !id.isEmpty())
-                .map(getResource(realmName)::findByClientId)
-                .filter(clients -> !clients.isEmpty())
-                .map(clients -> clients.get(0));
+        List<ClientRepresentation> foundClients = getResource(realmName).findByClientId(Objects.requireNonNull(clientId));
+
+        Optional<ClientRepresentation> client;
+        if (foundClients.isEmpty()) {
+            client = Optional.empty();
+        } else {
+            client = Optional.of(foundClients.get(0));
+        }
+
+        return client;
+    }
+
+    public Optional<ClientRepresentation> searchByName(String realmName, String name) {
+        // this is expensive, but easy to implement.
+        // if this too expensive, please provide a PR which implement a pagination for findAll()
+        Optional<ClientRepresentation> foundClients = realmRepository.getResource(realmName)
+                .partialExport(false, true)
+                .getClients()
+                .stream()
+                .filter(client -> Objects.equals(name, client.getName()))
```

Comment:
Hi @jkroepke 

It could be that the proposed implementation of the method is not deterministic enough. The client name is an optional value in the Keycloak data model and can be null. In that case, if the value of the name parameter is also null, the method will pick up the first client for which the name had not been defined.

And it looks like nothing prevents that scenario, please have a look at the `getClientIdentifier`method:

```java
    private String getClientIdentifier(ClientRepresentation client) {
        return client.getClientId() != null ? client.getClientId() : client.getName();
    }
``` 

Please consider adding the `Objects.requireNonNull(name)` here as well to keep the behaviour of the method similar to the `searchByClientId`

Reply:
Hi! Thanks for a review!

The `getClientIdentifier` is only used for logs.

There is a check, if name is not null.

https://github.com/adorsys/keycloak-config-cli/blob/1ed87218cdf281094b6af7dc8b798d51a73f4652/src/main/java/de/adorsys/keycloak/config/service/ClientImportService.java#L172-L178

But I will add a require NonNull here.